### PR TITLE
Remove encoding routine detritus and support encoding NULL values.

### PR DESCRIPTION
**Pull request by petermattis**

### CODE REVIEW

**Comment by petermattis** [comment]
The full diff is a bit overwhelming. The individual commits are decently self-contained.

**Comment by tamird** [comment]
LGTM.

File: sql/scan.go [hunk 1]
```diff
@@ -79,7 +79,9 @@ type scanNode struct {
 	kvIndex          int               // current index into the key/value pairs
 	rowIndex         int               // the index of the current row
 	colID            ColumnID          // column ID of the current key
+	valTypes         []parser.Datum    // the index key value types for the current row
 	vals             []parser.Datum    // the index key values for the current row
+	implicitValTypes []parser.Datum    // the implicit value types for unique indexes
```
Comment (tamird) [inline sql/scan.go]: a comment about these ...Types members being necessary to handle NULL's lack of type-preservation might be good

Comment (petermattis) [inline sql/scan.go]: I'm going to add a comment to `decodeKeyVals` as that method is the driver for these members.

Comment (tamird) [inline sql/scan.go]: SGTM

**Pull request merged**